File: packages/governance-py/src/aumos_governance/consent/manager.py
# SPDX-License-Identifier: BSL-1.1
# Copyright (c) 2026 MuVeraAI Corporation
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel

from aumos_governance.config import ConsentConfig
from aumos_governance.consent.store import ConsentRecord, ConsentStore
from aumos_governance.errors import ConsentNotFoundError


class ConsentCheckResult(BaseModel, frozen=True):
    """
    Result of a consent check.

    Attributes:
        granted: True if valid consent exists for the request.
        agent_id: The agent that was checked.
        data_type: The data type that was checked.
        purpose: The purpose that was checked (may be None).
        reason: Human-readable explanation of the outcome.
        record: The matching :class:`~aumos_governance.consent.store.ConsentRecord`
            if consent was granted, else None.
    """

    granted: bool
    agent_id: str
    data_type: str
    purpose: str | None = None
    reason: str
    record: ConsentRecord | None = None


class ConsentManager:
    """
    Manages consent records for agent data access.

    Consent is always recorded explicitly by a human or trusted orchestrator.
    There is no proactive consent suggestion or inference.

    All data is stored in-memory. A new ConsentManager starts empty.

    Example::

        manager = ConsentManager()
        manager.record_consent(
            agent_id="agent-42",
            data_type="user_profile",
            purpose="personalisation",
            granted_by="admin@example.com",
        )
        result = manager.check_consent("agent-42", "user_profile", "personalisation")
        assert result.granted is True
    """

    def __init__(self, config: ConsentConfig | None = None) -> None:
        self._config = config or ConsentConfig()
        self._store = ConsentStore()

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def record_consent(
        self,
        agent_id: str,
        data_type: str,
        purpose: str | None,
        granted_by: str,
        expires_at: datetime | None = None,
    ) -> ConsentRecord:
        """
        Record explicit consent for an agent to access a data type.

        If a consent record for the same (agent_id, data_type, purpose)
        already exists it is replaced.

        Args:
            agent_id: The agent being granted access.
            data_type: The category of data being consented to.
            purpose: Optional purpose string further scoping the consent.
                Pass None to record blanket consent for all purposes.
            granted_by: Identifier of the human or system granting consent.
            expires_at: Optional UTC datetime after which consent expires.
                When None, consent does not expire automatically.

        Returns:
            The created :class:`~aumos_governance.consent.store.ConsentRecord`.

        Raises:
            ValueError: If ``agent_id``, ``data_type``, or ``granted_by``
                is an empty string.
        """
        if not agent_id:
            raise ValueError("agent_id must be a non-empty string.")
        if not data_type:
            raise ValueError("data_type must be a non-empty string.")
        if not granted_by:
            raise ValueError("granted_by must be a non-empty string.")

        record = ConsentRecord(
            agent_id=agent_id,
            data_type=data_type,
            purpose=purpose,
            granted_by=granted_by,
            expires_at=expires_at,
        )
        self._store.put(record)
        return record

    def check_consent(
        self,
        agent_id: str,
        data_type: str,
        purpose: str | None = None,
    ) -> ConsentCheckResult:
        """
        Check whether consent has been granted for a data access request.

        This is a read-only operation — it does not modify state.

        Args:
            agent_id: The agent requesting access.
            data_type: The type of data being accessed.
            purpose: The purpose for which access is needed.
                A blanket consent record (purpose=None) satisfies any
                purpose check.

        Returns:
            A :class:`ConsentCheckResult` describing the outcome.
        """
        record = self._store.find(
            agent_id=agent_id,
            data_type=data_type,
            purpose=purpose,
        )

        if record is not None:
            purpose_text = f" for purpose '{purpose}'" if purpose else ""
            return ConsentCheckResult(
                granted=True,
                agent_id=agent_id,
                data_type=data_type,
                purpose=purpose,
                reason=(
                    f"Consent granted for agent '{agent_id}' to access "
                    f"'{data_type}'{purpose_text} (granted by '{record.granted_by}')."
                ),
                record=record,
            )

        # No valid consent record found.
        if self._config.default_deny:
            purpose_text = f" for purpose '{purpose}'" if purpose else ""
            return ConsentCheckResult(
                granted=False,
                agent_id=agent_id,
                data_type=data_type,
                purpose=purpose,
                reason=(
                    f"No valid consent record found for agent '{agent_id}' "
                    f"accessing '{data_type}'{purpose_text}. "
                    "Defaulting to deny."
                ),
                record=None,
            )

        # Permissive mode — absence of a record means allow.
        purpose_text = f" for purpose '{purpose}'" if purpose else ""
        return ConsentCheckResult(
            granted=True,
            agent_id=agent_id,
            data_type=data_type,
            purpose=purpose,
            reason=(
                f"No explicit consent record for agent '{agent_id}' "
                f"accessing '{data_type}'{purpose_text}; "
                "permissive mode allows by default."
            ),
            record=None,
        )

    def revoke_consent(
        self,
        agent_id: str,
        data_type: str,
        purpose: str | None = None,
    ) -> None:
        """
        Revoke a consent record.

        Args:
            agent_id: The agent whose consent is being revoked.
            data_type: The data type for which consent is revoked.
            purpose: The purpose to revoke. Use None to revoke the blanket
                consent record (purpose=None). To revoke ALL records for an
                agent, call :meth:`revoke_all_for_agent`.

        Raises:
            ConsentNotFoundError: If no matching record exists to revoke.
        """
        removed = self._store.remove(
            agent_id=agent_id,
            data_type=data_type,
            purpose=purpose,
        )
        if not removed:
            raise ConsentNotFoundError(
                agent_id=agent_id,
                data_type=data_type,
                purpose=purpose,
            )

    def revoke_all_for_agent(self, agent_id: str) -> int:
        """
        Revoke all consent records for an agent.

        Args:
            agent_id: The agent whose records should all be removed.

        Returns:
            The number of records revoked.
        """
        return self._store.remove_all_for_agent(agent_id)

    def list_consents(self, agent_id: str) -> list[ConsentRecord]:
        """
        Return all consent records for an agent.

        Both active and expired records are returned. Use
        :meth:`~aumos_governance.consent.store.ConsentRecord.is_expired`
        to filter if needed.

        Args:
            agent_id: The agent ID to query.

        Returns:
            List of :class:`~aumos_governance.consent.store.ConsentRecord` objects.
        """
        return self._store.list_for_agent(agent_id)

File: packages/governance-ts/src/consent/manager.ts
// SPDX-License-Identifier: BSL-1.1
// Copyright (c) 2026 MuVeraAI Corporation

import type { ConsentRecord } from '../types.js';
import type { ConsentConfig } from '../config.js';
import { parseConsentConfig } from '../config.js';
import { ConsentStore } from './store.js';

/** Options accepted by ConsentManager.recordConsent(). */
export interface RecordConsentOptions {
  /** ISO 8601 datetime after which the consent expires. */
  expiresAt?: string;
}

/** Result returned by ConsentManager.checkConsent(). */
export interface ConsentCheckResult {
  /** Whether valid consent exists. */
  readonly permitted: boolean;
  /** Matching consent record, if one was found. */
  readonly record?: ConsentRecord;
  /** Human-readable explanation on denial. */
  readonly reason?: string;
}

/**
 * ConsentManager records, checks, and revokes consent grants.
 *
 * Consent is a record of agreement by a principal (`grantedBy`) that a
 * specific agent may access a data type for a stated purpose.  Consent
 * is never inferred, suggested proactively, or upgraded automatically.
 *
 * Storage is delegated to ConsentStore (in-memory, Map-based).  Records
 * are never physically deleted; revocation sets `active: false`.
 *
 * Public API (Fire Line — do NOT add methods beyond these three core ones):
 *   recordConsent()  — persist a new consent grant
 *   checkConsent()   — check whether valid consent exists
 *   revokeConsent()  — revoke one or more consent records
 *
 * Additional read-only helper:
 *   listConsents()   — enumerate active consents for an agent
 */
export class ConsentManager {
  readonly #config: ConsentConfig;
  readonly #store: ConsentStore;

  constructor(config: unknown = {}) {
    this.#config = parseConsentConfig(config);
    this.#store = new ConsentStore();
  }

  // ---------------------------------------------------------------------------
  // Public API
  // ---------------------------------------------------------------------------

  /**
   * Records a consent grant for an agent to access a data type for a purpose.
   *
   * When an unexpired active record already exists for the same
   * (agentId, dataType, purpose) triplet, an additional record is still
   * created — multiple overlapping consents are valid and tracked separately.
   *
   * @param agentId   - The agent being granted consent.
   * @param dataType  - The category of data the agent may access.
   * @param purpose   - The stated purpose for which access is granted.
   * @param grantedBy - The principal authorising the consent.
   * @param options   - Additional options (e.g., expiry).
   * @returns The newly created ConsentRecord.
   */
  recordConsent(
    agentId: string,
    dataType: string,
    purpose: string,
    grantedBy: string,
    options: RecordConsentOptions = {},
  ): ConsentRecord {
    if (agentId.trim().length === 0) {
      throw new RangeError('agentId must be a non-empty string.');
    }
    if (dataType.trim().length === 0) {
      throw new RangeError('dataType must be a non-empty string.');
    }
    if (purpose.trim().length === 0) {
      throw new RangeError('purpose must be a non-empty string.');
    }
    if (grantedBy.trim().length === 0) {
      throw new RangeError('grantedBy must be a non-empty string.');
    }

    const record: ConsentRecord = {
      id: crypto.randomUUID(),
      agentId,
      dataType,
      purpose,
      grantedBy,
      grantedAt: new Date().toISOString(),
      expiresAt: options.expiresAt,
      active: true,
    };

    this.#store.add(record);
    return record;
  }

  /**
   * Checks whether valid consent exists for an agent to access a data type.
   *
   * When `purpose` is provided, only records with a matching purpose are
   * considered.  When omitted, any active record for the data type is accepted.
   *
   * Consent is also implicitly granted when:
   * - `requireConsent` is false in the config (default), AND
   *   no explicit record is required.
   * - The stated `purpose` is in the config's `defaultPurposes` list.
   *
   * @param agentId  - The agent to evaluate.
   * @param dataType - The data type being accessed.
   * @param purpose  - Optional purpose to match against.
   */
  checkConsent(agentId: string, dataType: string, purpose?: string): ConsentCheckResult {
    // Check whether this purpose is in the default (always-accepted) list.
    if (purpose !== undefined && this.#config.defaultPurposes !== undefined) {
      if (this.#config.defaultPurposes.includes(purpose)) {
        return { permitted: true };
      }
    }

    // When requireConsent is false and no explicit record is needed,
    // permit without checking the store.
    if (!this.#config.requireConsent) {
      // Still check for an explicit record — if one exists, surface it.
      const records = this.#store.findActive(agentId, dataType, purpose);
      const record = records[0];
      return { permitted: true, record };
    }

    // requireConsent is true — must find an active record.
    const records = this.#store.findActive(agentId, dataType, purpose);
    if (records.length === 0) {
      const purposeClause = purpose !== undefined ? ` for purpose "${purpose}"` : '';
      return {
        permitted: false,
        reason:
          `No active consent found for agent "${agentId}" ` +
          `to access data type "${dataType}"${purposeClause}.`,
      };
    }

    return { permitted: true, record: records[0] };
  }

  /**
   * Revokes consent for an agent to access a data type.
   *
   * When `purpose` is provided, only records with that exact purpose are
   * revoked.  When omitted, all active records for the data type are revoked.
   *
   * @param agentId  - The agent whose consent is being revoked.
   * @param dataType - The data type to revoke access to.
   * @param purpose  - Optional purpose to narrow the revocation scope.
   * @returns The number of records that were revoked.
   */
  revokeConsent(agentId: string, dataType: string, purpose?: string): number {
    return this.#store.revoke(agentId, dataType, purpose);
  }

  /**
   * Returns all currently active (non-revoked, non-expired) consent records
   * for an agent.
   */
  listConsents(agentId: string): readonly ConsentRecord[] {
    return this.#store.getActive(agentId, new Date());
  }
}
